pypose/lietensor/lietensor.py

```diff
@@ -29,100 +33,103 @@
                      'index_select', 'masked_select', 'index_copy', 'index_copy_',
                      'select', 'select_scatter', 'index_put','index_put_', 'copy_']
 
-class LieType:
+class LieType(ABC):
     '''LieTensor Type Base Class'''
     def __init__(self, dimension, embedding, manifold):
         self._dimension = torch.Size([dimension]) # Data dimension
         self._embedding = torch.Size([embedding]) # Embedding dimension
         self._manifold  = torch.Size([manifold])  # Manifold dimension
 
     @property
-    def dimension(self):
+    def dimension(self) -> torch.Size:
         return self._dimension
 
     @property
-    def embedding(self):
+    def embedding(self) -> torch.Size:
         return self._embedding
 
     @property
-    def manifold(self):
+    def manifold(self) -> torch.Size:
         return self._manifold
 
     @property
-    def on_manifold(self):
+    def on_manifold(self) -> bool:
         return self.dimension == self.manifold
 
-    def add_(self, input, other):
+    def add_(self, input: LieTensor, other: LieTensor) -> torch.Tensor:
         if self.on_manifold:
             other1 = torch.Tensor.as_subclass(input, torch.Tensor)
             other2 = torch.Tensor.as_subclass(other, torch.Tensor)
             return input.copy_(other1 + other2[..., :self.manifold[0]])
         raise NotImplementedError("Instance has no add_ attribute.")
 
-    def Log(self, X):
+    def Log(self, X: LieTensor | torch.Tensor) -> LieTensor:
```

Comment:
I see a lot of `LieTensor | torch.Tensor`, but I think not all functions allow `torch.Tensor` as input. For example, I remember this `Log` function may be one of them, it only allows Lie group LieTensor.

Reply:
A line like `x = x.tensor() if isinstance(x, LieTensor) else x` is to tolerate unintentional inputs, and even if that happens, the program doesn't raise an error without telling any cues. But as a typing annotation, we should not tell the users it is designed to accept `Tensor`.
 However, I believe some functions are accepting both `Tensor` and `LieTensor`, we need to be careful about that.